### ISSUE TITLE
Third-party cookies are going to be blocked by most browsers

### ISSUE THREAD

**Issue by hadimc** [ForgeRock/oidcSessionCheck#14]
Has there been any thoughts/conversations around how the library can be used if application sits on a different domain, and hence the browser (currently Safari and Chrome-Incognito) may block third party cookie send by the oidcSessionCheck iFrame?

**Comment by jakefeasel** [comment]
Yes, I've been investigating this particular issue quite a lot recently. So far, I believe the best course of action will be to implement the [Storage Access API](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API) with some feature detection. I don't have an estimate for exactly when I will be able to add this, but it's definitely on my radar.

**Comment by hadimc** [comment]
Thanks @jakefeasel, that looks promising. While I am a fan of it, I would like to suggest an alternative approach for checking FR sessions in an OIDC integration scenario, if I may. 

The main assumption I am making is that "_**ForgeRock is capable of relating an OAuth2 access token to a user session that has authorized the access**_". If this holds, let's assume

- ForgeRock can be the OAuth2 authz server as well as the resource server for both "identity" (via OIDC) and "session validation" (e.g., fr:session scope).
-  A client who has been granted an access token with "openid fr:session ..." scopes can use the token to call a checkSession endpoint to validate the user's session related to this token.

IMHO, an advantage of this approach is that it is built on an existing OAuth2/OIDC handshake and follows a standard SSO mechanism without engaging cookie management over cross-domain environments. Would love to hear your feedback.

**Comment by jakefeasel** [comment]
I actually think it would be more reasonable to just use the id_token to call out to the AS to check the session, rather than introducing a "special" access token. This would be more aligned with some of the other, current OIDC session management specs (the end session endpoint takes the id_token as the "id_token_hint" parameter, for example).

I can picture how this might work today as a ForgeRock-specific implementation. Unfortunately, one of the design goals I have with this library is that I want it to be completely standards-based, and as it stands today the standards groups haven't come up with an agreed-upon approach for this. That being said - maybe we can open up some talks with the OpenID Connect working group on the subject. I may try out a PoC with this library as the starting point.

**Comment by jakefeasel** [comment]
Quick update on this one - the Storage Access API won't work here, since it is designed for interactive cross-domain iframe elements, not hidden iframes. I honestly think we are in need of an updated rfc for OIDC Session Management, in light of all of this.

I actually have a proposal for exactly that that I'm working on getting in front of the openid spec working group. In the meantime, I have a few other ideas for how we can deal with this problem. I'll be writing an article soon that goes into detail about what I think we can do about this, today. I'll post a follow up to this issue when that article is written.

**Comment by hadimc** [comment]
Thanks for the update Jake.

**Comment by andycory-kcom** [comment]
I'm very interested in your potential solutions, @jakefeasel - I've been using your library in a PoC for a customer - worked great in Chrome but not in Safari. The Storage Access API seemed like the silver bullet until I also noticed the interactive requirement :-(

**Comment by jakefeasel** [comment]
Please see version 1.1.0, and the updated [README section on non-standard options](https://github.com/ForgeRock/oidcSessionCheck#non-standard-options-available-to-trusted-clients). Assuming you are using ForgeRock as your IDP, and assuming you "trust" your RPs, you have an option to use this library in a way that avoids cookie blocking behavior in the browser. I hope that helps you with this issue. I'll leave this issue open for now, since I really want to get a standards-based approach available (and one that doesn't require trusting the RP so completely). Maybe someday the working groups will draft something in response to this problem.
